### PR TITLE
Remove NodeInfo for csiNode from cache when there is no Pod scheduled

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This is related to #77595
In schedulerCache#RemoveCSINode, when there is no Pod on the node, we should remove the corresponding NodeInfo.
Without this removal, some NodeInfo's with nil CSINode may be stuck in the cache.

```release-note
NONE
```

### CODE REVIEW

**Comment by tedyu** [comment]
@msau42 @bsalamat @bertinatto 
Please review

File: pkg/scheduler/internal/cache/cache.go [hunk 1]
```diff
@@ -606,8 +606,12 @@ func (cache *schedulerCache) RemoveCSINode(csiNode *storagev1beta1.CSINode) erro
 	if !ok {
 		return fmt.Errorf("node %v is not found", csiNode.Name)
 	}
-	n.info.SetCSINode(nil)
-	cache.moveNodeInfoToHead(csiNode.Name)
+	if len(n.info.Pods()) == 0 && n.info.Node() == nil {
+		cache.removeNodeInfoFromList(csiNode.Name)
```
Comment (bsalamat) [inline pkg/scheduler/internal/cache/cache.go]: I am not sure if we should remove NodeInfo here. What is the scenario that we are addressing here?
I would expect to receive a separate event for removal of a node where the event handler can remove nodeInfo from the list and from the cache.

Comment (msau42) [reply pkg/scheduler/internal/cache/cache.go]: The Add/Update handlers for CSINode above will create a NodeInfo entry if one doesn't exist, so I think this is for cleaning that up.

I guess the question is, is there a scenario where CSINode object will be created/updated if a Node object doesn't exist?

Comment (tedyu) [reply pkg/scheduler/internal/cache/cache.go]: There is similar check on line 562.
I wonder if the condition there needs to consider the presence of CSINode.

Comment (bsalamat) [reply pkg/scheduler/internal/cache/cache.go]: > I guess the question is, is there a scenario where CSINode object will be created/updated if a Node object doesn't exist?

This is my question too. On top of that, even if a CSINode triggers a NodeInfo object creation, deletion of it should not necessarily delete NodeInfo object. I know there are checks to ensure that Node() is nil and node does not have any pods, but we should not add logic/code without having a good reason.

Comment (tedyu) [reply pkg/scheduler/internal/cache/cache.go]:
```
deletion of it should not necessarily delete NodeInfo object.
```
Wouldn't the remaining NodeInfo object form memory leak (without CSINode) ?

Comment (bertinatto) [reply pkg/scheduler/internal/cache/cache.go]: My understanding was that NodeInfo should exist for as long as the Node object exists, and the best way to guarantee that is to only delete it when Node is deleted. I'm not aware of scenarios where a CSINode might be created/updated if a Node doesn't exist.

Comment (tedyu) [reply pkg/scheduler/internal/cache/cache.go]: @bertinatto 
For the check on line 562:
```
	if len(n.info.Pods()) == 0 && n.info.Node() == nil {
		cache.removeNodeInfoFromList(node.Name)
```
Should we verify that n.csiNode is nil ?

Comment (tedyu) [reply pkg/scheduler/internal/cache/cache.go]: @bertinatto
Another point I don't fully understand is the following in schedulerCache#RemoveCSINode
```
	cache.moveNodeInfoToHead(csiNode.Name)
```
Why moving the node info to the head ?

Comment (bsalamat) [reply pkg/scheduler/internal/cache/cache.go]: > My understanding was that NodeInfo should exist for as long as the Node object exists, and the best way to guarantee that is to only delete it when Node is deleted. I'm not aware of scenarios where a CSINode might be created/updated if a Node doesn't exist.

@bertinatto In that case, should we delete lines 579, 580, 593, and 594 and instead throw an error in case we received a CSINode event for a node that does not exist? One thing to consider is that if a node creation and a CSINode creation events are sent at the same time, there could be race and we should have logic to handle the case that one may arrive before the other. The logic will be the same as what we have now, i.e. we should keep the 4 lines, but if we never expect a CSINode to arrive almost at the same time, we may want to consider removing the 4 lines and throw an error instead.

> Another point I don't fully understand is the following in schedulerCache#RemoveCSINode
	cache.moveNodeInfoToHead(csiNode.Name)
 Why moving the node info to the head ?

@tedyu whenever there is a change to a NodeInfo, we should move it to the head so that it is snapshotted in the next scheduling cycle. Removal of a CSINode from a NodeInfo is a change to the object. So, moving it to the head is correct.

Comment (msau42) [reply pkg/scheduler/internal/cache/cache.go]: When csi migration is turned on, we will create CSINode object during kubelet initialization.  Now, CSINode object creation requires Node object already exists (it has an OwnerRef to Node), but there's still the possibility that the informers can handle them out of order.

Comment (tedyu) [reply pkg/scheduler/internal/cache/cache.go]: @msau42 @bsalamat 
It seems recent discussion resolves around the timing of addition / removal of Node and CSINode.
How about adding some log for the !ok case so that we would know the timing in future troubleshooting ?

Comment (bertinatto) [reply pkg/scheduler/internal/cache/cache.go]: Currently the kubelet requires an existing `Node` before creating the `CSINode`:

https://github.com/kubernetes/kubernetes/blob/003c4e51555f6e9fce8b62ba6916d5b70b2f3010/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L444-L464

However, as @msau42 pointed out, if the informer handled the objects out of order *or* something else creates/deletes random `CSINode` objects (without ownref), nothing will remove `NodeInfo` from the list.

I think we could address this by:

a) Doing what @bsalamat suggested, i.e., throw an error in case we receive a CSINode event for a node that does not exist.
b) Doing what @tedyu is proposing in this PR.

I'd be in favor of #a, but if the informers can really handle objects out of order, we might end up with a `NodeInfo` without the volume limits (when `CSINode` is informed before `Node`).

In that case I think it makes sense to go with #b, as long as we make sure the `Node` doesn't exist.

Does this make sense?

Comment (tedyu) [reply pkg/scheduler/internal/cache/cache.go]:
```
as long as we make sure the Node doesn't exist.
```
@bertinatto 
Just want to confirm that the following does what is suggested above:
```
       if len(n.info.Pods()) == 0 && n.info.Node() == nil {
```

Comment (msau42) [reply pkg/scheduler/internal/cache/cache.go]: Regarding a), throwing an error if Node doesn't exist means that when the Node finally is added, we will miss updating CSINode.  So I think b) is better, but we need to think carefully through all the different scenarios to make sure we don't leak.

Comment (tedyu) [reply pkg/scheduler/internal/cache/cache.go]: I added log in the RemoveNode and RemoveCSINode funcs

Comment (bertinatto) [reply pkg/scheduler/internal/cache/cache.go]: I played with different scenarios and I think the change proposed in this PR is OK:

Scenario: AddNode, AddCSINode, RemoveCSINode, RemoveNode
1) NodeInfo is added to cache.nodes AND NodeInfo.Node is set to new Node
2) NodeInfo.CSINode is set to new CSINode
3) NodeInfo.CSINode is set to nil, but NodeInfo is still in the cache.nodes
4) NodeInfo is removed from cache.nodes

Scenario: AddCSINode, AddNode, RemoveCSINode, RemoveNode
1) NodeInfo is added to cache.nodes AND NodeInfo.CSINode is set to new CSINode
2) NodeInfo.Node is set to new Node
3) NodeInfo.CSINode is set to nil, but NodeInfo is still in the cache.nodes
4) NodeInfo is removed from cache.nodes

Scenario: AddNode, AddCSINode, RemoveNode, RemoveCSINode
1) NodeInfo is added to cache.nodes AND NodeInfo.Node is set to new Node
2) NodeInfo.CSINode is set to new CSINode
3) NodeInfo is removed from cache.nodes
4) Returns error because NodeInfo is not found in cache.nodes

Scenario: AddNode, RemoveNode, AddCSINode, RemoveCSINode
1) NodeInfo is added to cache.nodes AND NodeInfo.Node is set to new Node
2) NodeInfo is removed from cache.nodes
3) NodeInfo is added to cache.nodes AND CSINode is set in NodeInfo.CSINode
4) NodeInfo is removed from cache.nodes because NodeInfo.Node == nil 

As an aside, we should keep checking `len(n.info.Pods()) == 0` because AddPod also relies on the NodeInfo object.

File: pkg/registry/storage/csinode/strategy.go [hunk 2]
```diff
@@ -75,9 +75,9 @@ func (csiNodeStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Ob
 	newCSINode := obj.(*storage.CSINode)
 	oldCSINode := old.(*storage.CSINode)
 
-	inUse := getAllocatablesInUse(oldCSINode)
-
 	if !utilfeature.DefaultFeatureGate.Enabled(features.AttachVolumeLimit) {
+		inUse := getAllocatablesInUse(oldCSINode)
+
```
Comment (bertinatto) [inline pkg/registry/storage/csinode/strategy.go]: Unrelated change

**Comment by tedyu** [comment]
@bertinatto 
I have addressed your comments.

File: pkg/scheduler/internal/cache/cache.go (hunk 1 above)
Comment (msau42) [reply pkg/scheduler/internal/cache/cache.go]: @bsalamat do you think we are breaking any assumption in the scheduler that if there's an entry in nodes cache, then a Node obj must exist?

**Comment by tedyu** [comment]
@msau42 @bsalamat 
Please review

**Comment by tedyu** [comment]
@bsalamat
Please review.

Thanks

File: pkg/scheduler/internal/cache/cache.go (hunk 1 above)
Comment (bsalamat) [reply pkg/scheduler/internal/cache/cache.go]: Cache has two structures that stores information about nodes: `cache.nodes[]` and `cache.nodeTree`. Items in the `cache.nodes[]` make a doubly linked list to track the nodes that are recently updated. The closer a node to the head of the linked list, the more recently updated it is. The scheduler uses this linked list to snapshot the nodes that are updated after the last snapshot.

In this PR, we are just deleting a node from `cache.nodes[]` and not from `NodeTree`. I guess the assumption is that when the scheduler receives a "Add Node" event, the node is added to the `NodeTree` and we expect the scheduler to receive a corresponding "Delete Node" event for it. Upon receipt of that event, NodeTree is cleaned-up.

The case that we are trying to solve here is that we only receive a "Add CSINode" and "Delete CSINode" events, but we don't see "Add Node" and "Delete Node" events.
To be honest, I am unhappy about this whole meddling of CSINode addition/deletion with the scheduler node cache. The sequence of events and cases that do not come up to mind right now could hurt us in the future. Bugs introduced by lack of clarity and understanding of these events would be very subtle and very time consuming. I think we should avoid change the whole logic of adding a node to the scheduler cache when CSINode event comes in and the corresponding node does not exist and remove this piece of code from here too. We should rethink about the problem and come up with a different solution altogether. If the assumption is that we should create a CSINode only after a Node exist, it might not be unreasonable to expect the node object to exists in the cache by the time the CSINode event arrives. If it doesn't we only log an error. Alternatively, we may want to consider some sort of retry/update to cover such scenarios. I understand that some solutions may seem harder to implement, but accumulation of logic similar to what we have here in the scheduler would lead to an over complicated piece of code. The scheduler in particular is one of most critical components of K8s. A cluster is unusable without it. So, it must be a very reliable. That's why I am not happy about the current solution.

**Comment by tedyu** [comment]
```
meddling of CSINode addition/deletion with the scheduler node cache
```
I agree with the above assessment.

@bsalamat 
If you agree, I can try taking out the code changes to scheduler node cache w.r.t. CSINode.
What do you think ?

File: pkg/scheduler/internal/cache/cache.go (hunk 1 above)
Comment (bertinatto) [reply pkg/scheduler/internal/cache/cache.go]: Thank you for the detailed explanation, @bsalamat. I see how this solution can be problematic in the future. Let me take a stab at this and try to come up with a better solution.

**Comment by bsalamat** [comment]
@tedyu 
> If you agree, I can try taking out the code changes to scheduler node cache w.r.t. CSINode.
What do you think ?

Just removing the changes is not enough. As @bertinatto mentioned, we should redesign the solution to this problem and come up with a mechanism that works with CSINode events well.
